plugins/smartvol/rest.go

```diff
@@ -0,0 +1,222 @@
+				Type:   b.Type,
+				PeerID: b.PeerID,
+				Path:   b.Path,
+			}
+		}
+	}
+
+	return reqVolCreate
+}
+
+func applyDefaults(req *smartvolapi.VolCreateReq) {
+	if req.SnapshotReserveFactor == 0 {
+		req.SnapshotReserveFactor = 1
+	}
+
+	// Snapshot reserve not required if not enabled
+	if !req.SnapshotEnabled {
+		req.SnapshotReserveFactor = 1
+	}
+}
+
+func smartVolumeCreateHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	var err error
+
+	var req smartvolapi.VolCreateReq
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusUnprocessableEntity, err)
+		return
+	}
+
+	if err := validateVolCreateReq(&req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, err)
+		return
+	}
+
+	_, err = volume.GetVolume(req.Name)
+	if err == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, gderrors.ErrVolExists)
+		return
+	} else if err != gderrors.ErrVolNotFound {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	applyDefaults(&req)
+
+	if req.SnapshotReserveFactor < 1 {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.New("Invalid Snapshot Reserve Factor"))
+		return
+	}
+
+	// Convert request to internal Structure
+	smartvolReq := smartvolapi.Volume{
+		Name:                    req.Name,
+		Transport:               req.Transport,
+		Force:                   req.Force,
+		Size:                    req.Size,
+		DistributeCount:         req.DistributeCount,
+		ReplicaCount:            req.ReplicaCount,
+		ArbiterCount:            req.ArbiterCount,
+		DisperseCount:           req.DisperseCount,
+		DisperseDataCount:       req.DisperseDataCount,
+		DisperseRedundancyCount: req.DisperseRedundancyCount,
+		SnapshotEnabled:         req.SnapshotEnabled,
+		SnapshotReserveFactor:   req.SnapshotReserveFactor,
+		LimitPeers:              req.LimitPeers,
+		LimitZones:              req.LimitZones,
+		ExcludePeers:            req.ExcludePeers,
+		ExcludeZones:            req.ExcludeZones,
+		SubvolZonesOverlap:      req.SubvolZonesOverlap,
+	}
+
+	if err := bricksplanner.PlanBricks(&smartvolReq); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Convert struct to traditional Volume Create format
+	reqVolCreate := getVolumeReq(&smartvolReq)
+
+	nodes, err := reqVolCreate.Nodes()
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, err)
+		return
+	}
+
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	lock, unlock, err := transaction.CreateLockSteps(req.Name)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc:   "vol-create.PrepareBricks",
+			UndoFunc: "vol-create.UndoPrepareBricks",
+			Nodes:    nodes,
+		},
+		{
+			DoFunc: "vol-create.CreateVolinfo",
```

Comment:
Though we still use the volume-create provided steps, which is nice, re-creating the volume-create steps here, which isn't really good IMO. If we go down this way, we'll need to keep tracking volume-create in this plugin.

We need to figure out a way to this better.